Assets/Scripts/Algorithms/Clustering/AgglomerativeClusterer.cs

```diff
@@ -52,50 +50,40 @@ public override void Cluster() {
       // Check whether the minimum distance exceeds the maximum cluster radius, in which case the
       // algorithm has converged. This produces a conservative solution because the radius of a
       // merged cluster is less than or equal to the sum of the original cluster radii.
-      if (minDistance >= _maxRadius) {
+      if (minDistance == Mathf.Infinity || minDistance > _maxRadius) {
         break;
       }
 
       // Check whether merging the two clusters would violate the size constraint.
-      if (_clusters[clusterIndex1].Size() + _clusters[clusterIndex2].Size() > _maxSize) {
+      if (clusters[clusterIndex1].Size + clusters[clusterIndex2].Size > _maxSize) {
         distances[clusterIndex1, clusterIndex2] = Mathf.Infinity;
         continue;
       }
 
-      // Merge the two clusters together.
-      int minClusterIndex = Mathf.Min(clusterIndex1, clusterIndex2);
-      int maxClusterIndex = Mathf.Max(clusterIndex1, clusterIndex2);
-      _clusters[minClusterIndex].Merge(_clusters[maxClusterIndex]);
-      _clusters[minClusterIndex].Recenter();
-      validClusterIndices.Remove(maxClusterIndex);
+      // Merge the two clusters together and keep the cluster with the smaller index, i.e.,
+      // clusterIndex2.
+      clusters[clusterIndex2].Merge(clusters[clusterIndex1]);
+      clusters[clusterIndex2].Recenter();
+      validClusterIndices.Remove(clusterIndex1);
 
       // Update the distances matrix using the distance between the cluster centroids.
       // TODO(titan): Change the distance metric to use average or maximum linkage.
-      for (int i = 0; i < minClusterIndex; ++i) {
-        if (distances[minClusterIndex, i] < Mathf.Infinity) {
-          distances[minClusterIndex, i] =
-              Vector3.Distance(_clusters[minClusterIndex].Coordinates, _clusters[i].Coordinates);
+      foreach (int i in validClusterIndices) {
+        float distance = Vector3.Distance(clusters[clusterIndex2].Centroid, clusters[i].Centroid);
+        if (i < clusterIndex2) {
+          distances[clusterIndex2, i] = distance;
+        } else if (i > clusterIndex2) {
+          distances[i, clusterIndex2] = distance;
         }
       }
-      for (int i = minClusterIndex + 1; i < _clusters.Count; ++i) {
-        if (distances[i, minClusterIndex] < Mathf.Infinity) {
-          distances[i, minClusterIndex] =
-              Vector3.Distance(_clusters[minClusterIndex].Coordinates, _clusters[i].Coordinates);
-        }
-      }
-      for (int i = 0; i < maxClusterIndex; ++i) {
-        distances[maxClusterIndex, i] = Mathf.Infinity;
-      }
-      for (int i = maxClusterIndex + 1; i < _clusters.Count; ++i) {
-        distances[i, maxClusterIndex] = Mathf.Infinity;
-      }
     }
```

Comment:
The distance matrix update iterates over all valid cluster indices, but doesn't it only need to update distances involving the merged cluster?
Can save some time with:
```cs
foreach (int i in validClusterIndices) {
  if (i == clusterIndex2) continue;

  float distance = Vector3.Distance(clusters[clusterIndex2].Centroid, clusters[i].Centroid);
  if (i < clusterIndex2) {
    distances[clusterIndex2, i] = distance;
  } else {  // i > clusterIndex2
    distances[i, clusterIndex2] = distance;
  }
}
```